lifecycle-hooks.md

```diff
@@ -0,0 +1,188 @@
+```bash
+kubectl apply -f yamls/clusters/docker-cluster-lifecycle-hooks.yaml
+```
+
+In the extension server you should see logs similar to:
+
+```bash
+...... other log lines
+I1014 02:50:50.860237       1 handlers.go:47] "BeforeClusterCreate is called"
+I1014 02:50:50.873831       1 handlers.go:165] "BeforeClusterCreate response is Success. retry: 0"
+I1014 02:51:38.782571       1 handlers.go:80] "AfterControlPlaneInitialized is called"
+```
+
+Note that the workload cluster could take a few minutes to fully come up. Each of the log lines above would take a few minutes in between to show up.
+
+### Delete the Cluster
+
+In this section we will delete the `docker-cluster-lifecycle-hooks` cluster and observe that the extension server is receiving and logging the event.
+
+> If you are feeling a adventurous try the [(Optional) Block Cluster deletion using Extension Server](#optional-block-cluster-deletion-using-extension-server) to block the deleting of a cluster using extension.
+
+To delete the cluster run:
+
+```bash
+kubectl delete cluster docker-cluster-lifecycle-hooks
+```
+
+In the extension server you should see logs similar to:
+
+```bash
+I1014 03:06:52.393668       1 handlers.go:131] "BeforeClusterDelete is called"
+I1014 03:06:52.408447       1 handlers.go:165] "BeforeClusterDelete response is Success. retry: 0"
+```
+
+### (Optional) Block Cluster deletion using Extension Server
+
+Till now we used a simple extension server that only logs the lifecycle events. What if we wanted to do more with the lifecycle hooks? In this section we will see how lifecycle hooks can be used to do more than just log events. 
+
+Let's change our extension server's behavior to send a "blocking response" when it receives a `BeforeClusterDelete` event.
+
+The extension server is pre-configured to always send a "allow response". Let's change this so that the extension server now send a "blocking response" to the `BeforeClusterDelete` event.
+
+```bash
+kubectl patch configmap docker-cluster-lifecycle-hooks-test-extension-hookresponses --patch-file yamls/extension/block-patch.yaml
+```
+
+Now, lets delete the cluster.
+
+```bash
+kubectl delete cluster docker-cluster-lifecycle-hooks
+```
+
+You should observe that the delete operation is blocked and in the extension server logs you should observe the following logs:
+
+```
+...... other log lines
+I1014 03:25:41.039429       1 handlers.go:131] "BeforeClusterDelete is called"
+I1014 03:25:41.046028       1 handlers.go:165] "BeforeClusterDelete response is Success. retry: 5"
+.....
+```
+
+Observe the `retry: 5` in the logs. The extension server is effectively telling the cluster to block the delete operation and re-check after 5 seconds. The management cluster calls the extension server every ~5 seconds till the extension send back a `retry: 0`.
+
+**Note:** DO NOT kill the terminal that is blocked in the `kubectl delete cluster docker-cluster-lifecycle-hooks` run.
```

Comment:
Why not? I don't think this has too much of an impact on performance. I think I'd prefer if this section was more clearly runnable in a single terminal window - all the directions to swtich and open terminals are confusing to me. I think it would be better to write it in an agnostic way and let users dictate their own flow.

Reply:
I think we don't have to stay stuck in kubectl delete cluster, right? The request initially sets the deletionTimestamp and after that it's just a watch. So we can just stop the cmd?  (I think there might also be a cmd where it just doesn't wait, while it doesn't change anything on the server-side, maybe --wait=false)

Reply:
But yeah maybe making this a single-terminal thing is easier for folks

Reply:
Agree that single-terminal would be a better flow. Updated the instruction so that the user can follow the entire doc in a single terminal.